File: Java/Remove-Nth-Node-From-End-of-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null){
            return null;
        }
        
        if (head.next == null){
            head = null;
            return head;      
        }        
        
        ListNode current = head;
        int c = 0;             
                
        while (current!=null){
            current = current.next;
            c += 1;
        }        
        if (c == n){
            head = head.next;
            return head;
        }    
                
        
        int t = 0;
        current = head;         
        while (current!=null){
            if (t==(c-n-1)){
                current.next = current.next.next;
            }
            current = current.next;
            t+=1;
        }
        return head;    
    }
}

File: Golang/Remove-Nth-Node-From-End-of-List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    if head.Next == nil && n == 1 {
        head = nil
        return head
    }
    
    size := 0
    current := head
    for current != nil {
        current = current.Next
        size++
    }
    
    if n == size {
        head = head.Next
        return head
    }
    
    if n == 1 {
        current := head
        for current.Next.Next != nil {
            current = current.Next
        }
        
        current.Next = nil
        return head
        
    }
    slow, fast := head, head
    
    for i := 0; i<n; i++ {
        fast = fast.Next
    }
    
    for fast.Next != nil {
        slow = slow.Next
        fast = fast.Next
    }
    
    slow.Next = slow.Next.Next
    
    return head
}
